src/main/java/me/coley/recaf/mapping/AsmMappingUtils.java

```diff
@@ -0,0 +1,161 @@
+                            + key + " gave more than 1 result: " + String.join(", " + classPreimages));
+                }
+                // if we have a preimage for the class, apply the mapping to that preimage class name
+                if (classPreimages.size() == 1) {
+                    String classPreimage = classPreimages.iterator().next();
+                    String memberName = key.substring(key.indexOf('.') + 1);
+                    key = classPreimage + "." + memberName;
+                }
+            }
+
+            // check if this class/member has been mapped before and transform the mapping accordingly
+            Collection<String> preimages = inverseMapping.get(key);
+            if (preimages.size() > 1) {
+                throw new IllegalStateException("Reverse mapping of " + key
+                        + " gave more than 1 result: " + String.join(", " + preimages));
+            }
+            if (preimages.size() == 1) {
+                preimageAwareUpdates.put(preimages.iterator().next(), value);
+            } else {
+                preimageAwareUpdates.put(key, value);
+            }
+        }
+
+        existing.putAll(preimageAwareUpdates);
+    }
+
+    /**
+     * Transforms a given mapping in ASM format (See
+     * {@link org.objectweb.asm.commons.SimpleRemapper#SimpleRemapper(Map)}) to a mapping where the values are in the
+     * key format of ASM mappings. This allows using the new value as keys for another transformation step.
+     *
+     * <p>For example the mapping {@code calc/MyCalculator.MAX_DEPTH -> MAX_DEPTH_LEVEL} will be transformed to
+     * {@code calc/MyCalculator.MAX_DEPTH -> calc/Calculator.MAX_DEPTH_LEVEL}
+     *
+     * @param mapping ASM mapping to transform.
+     * @return Transformed mapping where the value would be a valid key for another ASM transformation step.
+     */
+    public static Map<String, String> transformAsmMappingValuesToKeyFormat(Map<String, String> mapping) {
+        return mapping.entrySet().stream()
+                .map(AsmMappingUtils::transformSingleAsmMappingToKeyFormat)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private static Map.Entry<String, String> transformSingleAsmMappingToKeyFormat(Map.Entry<String, String> mapping) {
+        // Heavily inspired by SimpleRecordingRemapper.map()
+        String key = mapping.getKey();
+        String value = mapping.getValue();
+
+        // Don't map constructors/static-initializers
+        if (key.contains("<"))
+            return null;
+
+        boolean isMember = key.contains(".");
+        if (!isMember) {
+            // This is a class, just return the original mapping as its value is the applied value
+            return mapping;
+        }
+
+        // Don't map invokedynamic calls
+        boolean isInvokeDynamic = key.charAt(0) == '.';
```

Comment:
Just curious, what cases does this come up for?

Reply:
This is where my (somewhat limited) understanding of the remapping (and details of invokedynamic) ends. As stated in a comment earlier I was looking at `SimpleRecordingRemapper.map()` in order to not miss any case and that function is prepared to handle these. I did not want to just pretend the case can't exist.
(This is where the strict response to your comment stops and speculation starts)

Unfortunately I have no idea how this would be handled correctly. It certainly looks like it doesn't need to just for reading code. I suspect that the remapped classes in memory are very messed up right now. I did a small test with an interface and a lambda:

```java
@FunctionalInterface
public interface StringSupplier {
    String getString();
}
```
```java
public class Main {
    public static void main(String[] args) {
        StringSupplier supp = () -> "Hello!";
        System.out.println(supp.getString());
    }
}
```

I threw a jar with these two files into Recaf and renamed `StringSupplier.getString()` to `StringSupplier.getRenamedString()`. I then opened the disassembly for `main()`. This showed the following bytecode disassembly:
```
DEFINE PUBLIC STATIC main([Ljava/lang/String; args)V
ALIAS H_META "handle[H_INVOKESTATIC java/lang/invoke/LambdaMetafactory.metafactory(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodHandle;Ljava/lang/invoke/MethodType;)Ljava/lang/invoke/CallSite;]"
A:
LINE A 3
INVOKEDYNAMIC getString ()LStringSupplier; ${H_META} args[()Ljava/lang/String;, handle[H_INVOKESTATIC Main.lambda$main$0()Ljava/lang/String;], ()Ljava/lang/String;]
ASTORE supp
B:
LINE B 4
GETSTATIC java/lang/System.out Ljava/io/PrintStream;
ALOAD supp
INVOKEINTERFACE StringSupplier.getRenamedStr()Ljava/lang/String;
INVOKEVIRTUAL java/io/PrintStream.println(Ljava/lang/String;)V
C:
LINE C 5
RETURN
D:
```

To me it looks like the `LambdaMetafactory` will receive the old method name argument from `INVOKEDYNAMIC getString ()LStringSupplier;` while the invokeinterface instruction has been changed to the new name.
Again, this doesn't seem to bother the decompilers, but I guess this would be the case where remapping invokedynamic comes up.

I'll stress again that my understanding of the JVM and bytecode instructions is rather limited. This test was just a product of my curiosity.